### ISSUE TITLE
Loss functions without reduction

### ISSUE THREAD

**Issue by delta2323** [pfnet/chainer#2558]
We are now working on fixing loss functions so that they can output loss values *without reduction*. The objective of this issue is to manage the status of implementation. When we say a *loss function*, it means Chainer functions under the `chainer.functions.loss` namespace.

As discussed in #2325 We will take two-step strategy to support this feature. Specifically, we first add `reduce` option to loss functions that switches enable/disable reduction. The values the option take can vary depending on concrete loss functions. This change will be applied to current master branch (i.e. v1).
Next, we remove `reduce` option so that functions output loss function without reduction only. This change will merge to _v2 branch. Along this modification, `normalize` option in e.g.  `F.softmax_cross_entropy` is also removed. So this PR will also resolves #1565 . 

As there are many loss functions, **we appreciate it if you contribute to solve this issue**.

# Status

Note: We list all functions under the `chainer.functions.loss` namespace. We may not need to apply changes to some of them.

loss function name|v1|v2
-------------------|--|--
black_out|Done(#2600)|Working(#2682)
contrastive|Done(#2603)|Working(#2683)
crf1d|Done(#2559)|Working(#2684)
~~argmax_crf1d~~|Not needed|Not needed
cross_covariance|Done(#2697)|
connectionist_temporal_classification|Done(#2658)|
decov|Done(#2698)|
hinge|Done(#2577)|Working(#2706)
huber_loss|Done(#2560)|Working(#2707)
mean_absolute_error|Done(#2567)(*)|Working(#2563)
mean_squared_error|Done(#2566)(*)|Working(#2562)
negative_sampling|Done(#2704)|
sigmoid_cross_entropy|Done(#2325)|Working(#2561)
softmax_cross_entropy|Done(#2357)|Working(#2708)
triplet|Done(#2681)|Working(#2709)
gaussian_kl_divergence|Done(#2519)|Working(#2652)
bernoulli_nll|Done(#2525)|Working(#2653)
gaussian_nll|Done(#2526)|Working(#2654)

(*) We implemented `AbsoluteError` and `SquaredError`. The design of `MeanAbsoluteError` and `MeanSquaredError` is still under discussion.

**Comment by delta2323** [comment]
It could be helpful the implementation of #1223

**Comment by delta2323** [comment]
@unnonouno It seems we need not to change `crf1d` because the loss value is not of the form of summation of losses for each input instances and therefore we cannot decompose it. Is my understanding correct?

**Comment by delta2323** [comment]
Also, should we add `reduce` option to `argmax_crf1d`? `argmax_crf1d` computes loss values for each sample. I do not have an idea if it is appropriate operation that adds them up.

**Comment by unnonouno** [comment]
No. Loss function of crf1d is defined for a sequence. And, it gets a set of sequences and currently calculates mean of losses of sequences.

**Comment by unnonouno** [comment]
And `reduce` option is not required for `argmax_crf1d`.

**Comment by delta2323** [comment]
> And, it gets a set of sequences and currently calculates mean of losses of sequences.

Then, I thought we should add `reduce` option that takes `'no'` and `'mean'`. Do I misunderstand something?

**Comment by unnonouno** [comment]
No, that sounds good! And i'll make it.

**Comment by delta2323** [comment]
Thank you!

**Comment by Hakuyume** [comment]
For `MeanSquaredError` and `MeanAbsoluteError`, adding `reduce='no'` option looks bad to me because they have `Mean` in their names. The following solution looks good to me.

- Modify `MeanSquaredError`/`MeanAbsoluteError`
    - `reduce` option is for consistency and takes only `'mean'`.
    - These functions are removed in v2.
- Add `SquaredError`/`AbsoluteError`
    - `reduce` option takes `'mean'` or `'no'`.
    - If `reduce='mean'`, the behaviour is same as `MeanSquaredError`/`MeanAbsoluteError`.

How do you think of it?

**Comment by delta2323** [comment]
I agree with you in that `reduce='no'` is odd for `Mean*Error`. But as an argument that takes only single value does is meaningless (except for keeping backward compatibility), I think we need not to introduce `reduce` if it takes only `'mean'`. Some user could need functions that compute *samplewise* loss (i.e. sum losses up along axes except the first one). If we would like to introduce it, `reduce` makes sense (e.g. it takes `mean_along_samples` and `samplewise`)

If users demand functions that calculate element-wise squared/absolute error, adding `SquaredError` and `AbsoluteError` seems good to me.

**Comment by Hakuyume** [comment]
@delta2323 Thank you. I want element-wise squared/absolute error. So, I made PRs for `SquaredError` and `AbsoluteError`.

**Comment by delta2323** [comment]
`F.cross_covariance` calculates the empirical covariance of `y` and `z` (after centering them) and  returns the (half of) the Frobenius norm of the covariance as a loss value. Does anyone know the appropriate output of the function "without reduction"?

**Comment by delta2323** [comment]
Same is true of `deconv`.

**Comment by beam2d** [comment]
I found that the default behavior of reduction in loss functions will be changed in these PRs for v2. I think these changes are too confusing and hard to follow for users (the output shape will be changed, maybe no error occurs on misuse). How do you think of it? Isn't it enough to just add `reduce` option and keep the default behavior as is?

**Comment by delta2323** [comment]
We discussed within the development team about the PRs for v2. To summarize, I think we need not remove `reduce` option that has been introduced in v1. So, all PRs submitted to v2 are no longer needed and should be closed.

The problem we wanted solve were as follows:
1. We sometimes want to element-wise loss values (e.g. loss functions related to VAE).
2. Inconsistency of the shape of output loss values. Although loss functions output scalar variables basically, `F.huber` outputs sample-wise loss values.
3. Inconsistency of semantics of loss values. Some loss functions take sum of loss values from each samples, whereas other functions take their mean.

The first problem is solved by introducing `reduce` option. To solve the inconsistency problems (2. and 3.), we tried to change that each loss function should output loss values *without reduction* by default. 
I understand this change should break users' code and that is why we send the PRs to v2. But I was also fearing that the effect of this change is so large, as pointed out by @beam2d in the previous comment. 

I realize that current `reduce` option solves 2. and 3. to some extent, although not perfectly. Users can align the shape of outputs if needed , which could a remedy for 2. and by looking at the value of `reduce`, users can be aware of the semantics of loss values, which is a partial answer for 3.

**Comment by Hakuyume** [comment]
I found an issue that some existing options are redundant. For example, `SigmoidCrossEntropy` has `normalize` option now.  Of course, we don't need it because we can set `reduce='mean_along_sample'` or `reduce='mean'`. However, we can't remove it without breaking backward compatibility. I think it is better to remove such options even if we break the backward compatibility because it makes the APIs more simple and it can be done by a tiny modification. How do you think of it?

**Comment by Hakuyume** [comment]
From my understanding, if I want a sigmoid cross entropy of 3-dimensional array which reduced along 0th and 2nd axis, I have to write `F.average(F.sigmoid_cross_entropy(x, t, reduce='mean_along_sample'), axis=1)'. Is it right?

**Comment by delta2323** [comment]
Currently `F.sigmoid_cross_entropy` does not support `mean_along_sample`, although we mention it in the context of `MeanSquaredError`. There are several choices:

* `F.average(F.average(F.sigmoid_cross_entorpy(x, t, reduce='no'), axis=0), axis=2)` (currently, `F.average` does not support reduction along multiple axes).
* `F.sum(F.sigmoid_cross_entropy(x, t, reduce='no')) / x.shape[0] / x.shape[2]`

**Comment by delta2323** [comment]
As @Hakuyume pointed out, the APIs of some loss functions are complicated and we can simplify them by removing `reduce` option. That is a good point of this change that I forgot to mention in [the comment above](https://github.com/pfnet/chainer/issues/2558#issuecomment-300378788).
`normalize` option in `F.sigmoid_cross_entropy` (as @Hakuyume mentioned), and `normalize`, `class_weight` in `F.softmax_cross_entropy` are such examples.

But we have not notified users these changes so much and we are not super confident. Further, we cannot evaluate to what extent this change causes problems. One thing I come up with as a a compromise (during writing this comment) is to issue a warning during v2 to notify the future changes, and remove `reduce` option in the next major version up.

**Comment by Hakuyume** [comment]
> Thank you for your comments. Currently F.sigmoid_cross_entropy does not support reduce='mean_along_sample', although we mention it in the context of MeanSquaredError. 

As you said, `reduce` option of `SigmoindCrossEntropy` doesn't support reduce='mean_along_sample'. However, we can support it without breaking backward compatibility. I just wanted to show an example in which `reduce` option is not enough and we have to use a reduction function together. Of course, this case is not so common.

> As @Hakuyume pointed out, the APIs of some loss functions are complicated and we can simplify them by removing reduce option.

I think it is better to remove `normalize` than to remove `reduce`, because most of functions have `reduce` option now and the effect of `normalize` is hard to understand. If we can remove one of them in the future, we should remove `normalize`. So the warning in `v2` should be for usage of `normalize`.

> ... class_weight in F.softmax_cross_entropy are such examples.

I think `class_weight` in `SoftmaxCrossEntropy` cannot be removed because it is not related to reduction. This is just a pointing out and doesn't affect this discussion.

**Comment by Hakuyume** [comment]
I agree with @beam2d that `reduce` option is enough to solve issues https://github.com/pfnet/chainer/issues/2558#issuecomment-299809571. But I have two concerns about @beam2d 's comment.

- the cost of implementing and testing a new loss will be increase.
    Supporting `reduce` option will increase the cost of implementing and testing. I think the code supporting `reduce` option is more complicated than that of elementwise version.
- current behaviour is sometimes difficult.
    @beam2d suggested
    > Isn't it enough to just add reduce option and keep the default behavior as is?

    However, I think some default behaviours are hard to understand for users. For example, `HuberLoss` https://github.com/pfnet/chainer/issues/2415. Perhaps this is not a big problem because user can know the default behaviour by checking the default value of `reduce`.

**Comment by delta2323** [comment]
> I just wanted to show an example in which reduce option is not enough and we have to use a reduction function together.

OK, I understand it. I think it is OK to support `reduce='mean_along_sample'` or `'some_along_sample'` (or something like that).

> I think class_weight in SoftmaxCrossEntropy cannot be removed because it is not related to reduction. This is just a pointing out and doesn't affect this discussion.

Sorry, it was my misunderstanding.

**Comment by delta2323** [comment]
> the cost of implementing and testing a new loss will be increase

One possible solution is to make "reduce" option accept a single value (e.g. "no") and leave a room for widening the support of other values. If the option that takes single value is weird, another choice is not to provide "reduce" option and to write how the loss function reduces the loss values explicitly.

**Comment by beam2d** [comment]
> the cost of implementing and testing a new loss will be increase.
I understand this point, but we should be very careful on breaking user codes. If we cannot suggest the way to fix their code when they run the old code, I strongly recommend you not to make such a change even if it will reduce the development cost.

> current behaviour is sometimes difficult.
This discussion then changes the balance between taking the development cost vs. not (or safely) breaking user codes. If the current behavior is unnatural, we have to change the default behavior because users will write wrong codes and then be confused by the strange behavior. I want to let them know the correct way to use such a feature when they misuse it.

As a consequence, I propose to change the default behavior of these functions whose default value is strange (including huber_loss). This should follow the standard way of Python to change the existing behavior, i.e., let the use of default value raise `FutureWarning`, and then change that behavior in a future version. The version in which `FutureWarning` is raised must be earlier than the version in which the behavior is actually changed. How do you guys think of it?

**Comment by delta2323** [comment]
Although the discussion is continuing, as the release date of the v2 is close, I close this issue with [my proposal above](https://github.com/pfnet/chainer/issues/2558#issuecomment-300378788) as a tentative conclusion. We welcome opinions and discussions for the improvement in the future versions and thank all who involved in the discussion so far.
